### PR TITLE
Introduce go.opentelemetry.io/otel/internal/env

### LINKED ISSUE #3766: Refactor env var parsing
We have similar (sometimes the same) env var parsing in a few places:

- https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/jaeger/env.go
- https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/zipkin/env.go
- https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/internal/env/env.go

I am not convinced that if we want have one package with all possible env vars like https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/internal/env/env.go. I feel such package would not be "cohesive". I would prefer to have a package with code that helps parsing env vars. We should define the env var keys and use these helper functions in packages that need these env vars.

I propose to create the env var parsing code in `go.opentelemetry.io/otel/internal/env`. The reason is that `go.opentelemetry.io/otel/exporters/*` does not have access to the current `go.opentelemetry.io/otel/sdk/internal/env `.

The package could have following functions (I do not think we need more):

```go
 func String(defaultValue string, keys ...string) string // for endoints, names
 func Duration(defaultValue time.Duration, keys ...string) time.Duration  // for timeouts and intervals
 func Int(defaultValue int, keys ...string) int // for limits
```

PS. If anything is not clear than maybe creating a PR will be easier for me 😅

### PR DESCRIPTION
**Pull request by pellared**
## Why

Fixes #3766

Supersedes #3764

> The SDK MUST interpret an empty value of an environment variable the same way as when the variable is unset.

Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#parsing-empty-value

## What

- Introduce `go.opentelemetry.io/otel/internal/env` package to centrilize the env var parsing code
- Handle empty env vars as it they were not set (this was not always done and this is [required by the OTel spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#parsing-empty-value))
- Fallback to "next" env vars in case the previous are invalid (this is just a refinement of existing limits env vars parsing)
- Remove `go.opentelemetry.io/otel/sdk/internal/env` package

Side note: OTLP packages have different env var parsing logic (they use `EnvOptionsReader`). I do not want to touch it (for sure not in scope in this PR).

### CODE REVIEW

File: sdk/trace/env.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+// Environment variable names.
+const (
+	// batchSpanProcessorScheduleDelayKey is the delay interval between two
+	// consecutive exports (i.e. 5000).
+	batchSpanProcessorScheduleDelayKey = "OTEL_BSP_SCHEDULE_DELAY"
+	// batchSpanProcessorExportTimeoutKey is the maximum allowed time to
+	// export data (i.e. 3000).
+	batchSpanProcessorExportTimeoutKey = "OTEL_BSP_EXPORT_TIMEOUT"
+	// BatchSpanProcessorMaxQueueSizeKey is the maximum queue size (i.e. 2048).
```
Comment (dmathieu) [inline sdk/trace/env.go:25]:
```suggestion
	// batchSpanProcessorMaxQueueSizeKey is the maximum queue size (i.e. 2048).
```

File: sdk/trace/env.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	batchSpanProcessorMaxQueueSizeKey = "OTEL_BSP_MAX_QUEUE_SIZE"
+	// BatchSpanProcessorMaxExportBatchSizeKey is the maximum batch size (i.e.
```
Comment (dmathieu) [inline sdk/trace/env.go:27]:
```suggestion
	// batchSpanProcessorMaxExportBatchSizeKey is the maximum batch size (i.e.
```

File: sdk/trace/env.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	// 512). Note: it must be less than or equal to
+	// EnvBatchSpanProcessorMaxQueueSize.
```
Comment (dmathieu) [inline sdk/trace/env.go:29]: There isn't any method by this name anymore. I think this should be

```suggestion
	// BatchSpanProcessorMaxQueueSize.
```

File: sdk/trace/env.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace // import "go.opentelemetry.io/otel/sdk/trace"
+
+// Environment variable names.
```
Comment (dmathieu) [inline sdk/trace/env.go:17]: Nit, but how about mentioning here that those constants are exposed to tests in `env_test.go`?

File: sdk/trace/env.go (hunk 3 above)
Comment (pellared) [inline sdk/trace/env.go:29]:
```suggestion
	// batchSpanProcessorMaxQueueSize.
```

File: sdk/trace/env.go (hunk 4 above)
Comment (pellared) [reply sdk/trace/env.go:17]: a20e39b4

**Review by dmathieu — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -50,7 +50,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Remove use of deprecated `"math/rand".Seed` in `go.opentelemetry.io/otel/example/prometheus`. (#3733)
 - Do not silently drop unknown schema data with `Parse` in  `go.opentelemetry.io/otel/schema/v1.1`. (#3743)
 - Data race issue in OTLP exporter retry mechanism. (#3756)
-- Fixes wrapping a nil error in some cases (#????)
+- Handle empty environment variable as it they were not set. (#3805)
```
Comment (pellared) [inline CHANGELOG.md:53]: Opportunistic fix 😉

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
I'm not comfortable with the consolidation on an internal package in the API module with dependence on that module by the SDK modules, as discussed [here](https://github.com/open-telemetry/opentelemetry-go/pull/3806#pullrequestreview-1315949613).  I think we need to find a solution that doesn't introduce compatibility risk by treating cross-module internal package imports as not part of the public API.

**Comment by Aneurysm9** [comment]
I think symlinks to duplicate code might work, but might be surprising and lead to unexpected consequences if someone thinks two files are independent when they're not.  It might also lead to issues with `porto` and canonical import path comments.

Code generation that produces duplicates and handles any minor changes between copies seems like the best approach I've heard thus far.  Doing that should also allow us to keep track of where duplicated code is being used.

Alternately, we can extract this into a separate module with a stable public API.  It might limit our flexibility going forward, but should be the least surprising in all use cases.

**Pull request closed without merging**